File: dev/python/automated_imgur_upload/gsheets2img.py
import json
import shutil
import tempfile
import zipfile
from pathlib import Path
import requests
from playwright.sync_api import sync_playwright

# Hardcoded constants
sheet_id = "1wWMIzaUKISAXMbOEnmsuuLkO9PesabpdTUWdosvHygM"
output_dir = Path(__file__).parent.parent.parent.parent / "public/images/equip_misc/"
include_sheets = []
exclude_sheets = [
    "(WiP) SS RLD Chart",
    "Copy of BB Guns",
    "(WiP) OpSi Image Guide Pt4",
    "Ammo Modifiers Chart",
    "(WiP) Rikka Specific Guide"
]
resolved_output_dir = output_dir.resolve()

def download(sheet_id):
    temp_dir = Path(tempfile.mkdtemp(prefix="gs2imgz-"))
    zip_path = temp_dir / f"{sheet_id}.zip"
    url = f"https://docs.google.com/spreadsheets/d/{sheet_id}/export?format=zip"
    response = requests.get(url)
    response.raise_for_status()
    with open(zip_path, "wb") as f:
        f.write(response.content)
    return zip_path

def unzip(zip_path):
    temp_dir = Path(tempfile.mkdtemp(prefix="gs2imgx-"))
    with zipfile.ZipFile(zip_path, "r") as zip_ref:
        zip_ref.extractall(temp_dir)
    shutil.rmtree(zip_path.parent, ignore_errors=True)
    return temp_dir

def screenshot(html_path, png_path, browser):
    page = browser.new_page(
        viewport={"width": 1920, "height": 1080}, device_scale_factor=2
    )
    page.goto(f"file://{html_path}", timeout=0)

    row_header = page.query_selector(".row-header-wrapper")
    if not row_header:
        print(f"Skipping {html_path.name}: row-header-wrapper not found.")
        page.close()
        return

    row_header_box = row_header.bounding_box()
    if not row_header_box:
        print(f"Skipping {html_path.name}: bounding box not found.")
        page.close()
        return

    tbody = page.query_selector("tbody")
    if not tbody:
        print(f"Skipping {html_path.name}: tbody not found.")
        page.close()
        return

    bounding_box = tbody.bounding_box()
    if not bounding_box:
        print(f"Skipping {html_path.name}: bounding box not found.")
        page.close()
        return

    row_header_width = row_header_box["width"]

    # Create a new clip area instead of modifying the original bounding_box
    clip_area = {
        "x": bounding_box["x"] + row_header_width + 1,
        "y": bounding_box["y"],
        "width": bounding_box["width"] - row_header_width - 1,
        "height": bounding_box["height"]
    }

    # Set viewport large enough to contain the full content with buffer
    page.set_viewport_size({
        "width": max(1920, int(clip_area["width"]) + 100),
        "height": max(1080, int(clip_area["height"]) + 100)
    })

    print(f"Uploading {html_path.stem}")
    page.screenshot(path=str(png_path), clip=clip_area)
    page.close()

def process_sheets(sheet_names, extracted_dir, browser):
    for sheet_name in sheet_names:
        file_name = sheet_name.replace(" ", "_") + ".jpeg"
        html_path = extracted_dir / f"{sheet_name}.html"
        png_path = resolved_output_dir / file_name
        if not html_path.exists():
            print(f"Skipping {sheet_name}: HTML file not found.")
            continue
        screenshot(html_path, png_path, browser)

def main():
    resolved_output_dir.mkdir(parents=True, exist_ok=True)
    zip_path = download(sheet_id)
    extracted_dir = unzip(zip_path)
    files = [f for f in extracted_dir.iterdir() if f.suffix == ".html"]
    sheet_names = [
        f.stem
        for f in files
        if (not include_sheets or f.stem in include_sheets)
        and f.stem not in exclude_sheets
    ]

    print(f"Found {len(sheet_names)} sheets to process.")

    with sync_playwright() as playwright:
        browser = playwright.firefox.launch()
        try:
            process_sheets(sheet_names, extracted_dir, browser)
        finally:
            browser.close()

    shutil.rmtree(extracted_dir)
    print("Finished uploading images.")

if __name__ == "__main__":
    main()

File: packages/gsheets2img/gsheets2img.ts
import fs from 'fs';
import path from 'path';
import tmp from 'tmp';
import AdmZip from 'adm-zip';
import { chromium } from 'playwright';

// Hardcoded constants
const sheetId = "1wWMIzaUKISAXMbOEnmsuuLkO9PesabpdTUWdosvHygM";
const outputDir = "../frontend/public/images/equip_misc/";
const includeSheets: string[] = [];
const excludeSheets = [
    "(WiP) SS RLD Chart",
    "Copy of BB Guns",
    "(WiP) OpSi Image Guide Pt4",
    "Ammo Modifiers Chart",
    "(WiP) Rikka Specific Guide"
];
const resolvedOutputDir = path.resolve(outputDir);

async function download(sheetId: string): Promise<string> {
    const tempDir = tmp.dirSync({ prefix: 'gs2imgz-' });
    const zipPath = path.join(tempDir.name, `${sheetId}.zip`);
    const url = `https://docs.google.com/spreadsheets/d/${sheetId}/export?format=zip`;
    const response = await fetch(url);
    if (!response.ok) {
        throw new Error(`Failed to download: ${response.statusText}`);
    }
    const buffer = await response.arrayBuffer();
    fs.writeFileSync(zipPath, Buffer.from(buffer));
    return zipPath;
}

function unzip(zipPath: string): string {
    const tempDir = tmp.dirSync({ prefix: 'gs2imgx-' });
    const zip = new AdmZip(zipPath);
    zip.extractAllTo(tempDir.name, true);
    fs.rmSync(path.dirname(zipPath), { recursive: true, force: true });
    return tempDir.name;
}

async function screenshot(htmlPath: string, pngPath: string, browser: any): Promise<void> {
    const page = await browser.newPage({
        viewport: { width: 1920, height: 1080 },
        deviceScaleFactor: 2
    });
    await page.goto(`file://${htmlPath}`, { timeout: 0 });

    const rowHeader = await page.$('.row-header-wrapper');
    if (!rowHeader) {
        console.log(`Skipping ${path.basename(htmlPath)}: row-header-wrapper not found.`);
        await page.close();
        return;
    }

    const rowHeaderBox = await rowHeader.boundingBox();
    if (!rowHeaderBox) {
        console.log(`Skipping ${path.basename(htmlPath)}: bounding box not found.`);
        await page.close();
        return;
    }

    const tbody = await page.$('tbody');
    if (!tbody) {
        console.log(`Skipping ${path.basename(htmlPath)}: tbody not found.`);
        await page.close();
        return;
    }

    const boundingBox = await tbody.boundingBox();
    if (!boundingBox) {
        console.log(`Skipping ${path.basename(htmlPath)}: bounding box not found.`);
        await page.close();
        return;
    }

    const rowHeaderWidth = rowHeaderBox.width;

    const clipArea = {
        x: boundingBox.x + rowHeaderWidth + 1,
        y: boundingBox.y,
        width: boundingBox.width - rowHeaderWidth - 1,
        height: boundingBox.height
    };

    await page.setViewportSize({
        width: Math.max(1920, Math.floor(clipArea.width) + 100),
        height: Math.max(1080, Math.floor(clipArea.height) + 100)
    });

    console.log(`Uploading ${path.parse(htmlPath).name}`);
    await page.screenshot({ path: pngPath, clip: clipArea });
    await page.close();
}

async function processSheets(sheetNames: string[], extractedDir: string, browser: any): Promise<void> {
    for (const sheetName of sheetNames) {
        const fileName = sheetName.replace(/ /g, '_') + '.jpeg';
        const htmlPath = path.join(extractedDir, `${sheetName}.html`);
        const pngPath = path.join(resolvedOutputDir, fileName);
        if (!fs.existsSync(htmlPath)) {
            console.log(`Skipping ${sheetName}: HTML file not found.`);
            continue;
        }
        await screenshot(htmlPath, pngPath, browser);
    }
}

async function main(): Promise<void> {
    fs.mkdirSync(resolvedOutputDir, { recursive: true });
    const zipPath = await download(sheetId);
    const extractedDir = unzip(zipPath);
    const files = fs.readdirSync(extractedDir).filter(f => f.endsWith('.html')).map(f => path.join(extractedDir, f));
    const sheetNames = files
        .map(f => path.parse(f).name)
        .filter(name => (!includeSheets.length || includeSheets.includes(name)) && !excludeSheets.includes(name));

    console.log(`Found ${sheetNames.length} sheets to process.`);

    const browser = await chromium.launch();
    try {
        await processSheets(sheetNames, extractedDir, browser);
    } finally {
        await browser.close();
    }

    fs.rmSync(extractedDir, { recursive: true, force: true });
    console.log("Finished uploading images.");
}

if (import.meta.url === `file://${process.argv[1]}`) {
    main().catch(console.error);
}